avahi-autoipd/main.c

```diff
@@ -1218,10 +1236,36 @@ static int loop(int iface, uint32_t addr) {
 
                 if (info.sender_ip_address == addr) {
 
+                    /* Normal conflict? */
                     if (memcmp(hw_address, info.sender_hw_address, ETHER_ADDRLEN)) {
-                        /* Normal conflict */
-                        conflict = 1;
-                        daemon_log(LOG_INFO, "Received conflicting normal ARP packet.");
+                        int should_defend = (state == STATE_RUNNING);
+
+                        /* Have we seen any other conflicting ARP packets within the last DEFEND_INTERVAL seconds? */
+                        if ((should_defend) && (has_timestamp(&last_conflict))) {
+                            /* Check if the time recorded for the previous conflicting ARP packet is recent */
+                            usec = avahi_age(&last_conflict) / 1000ULL;
+                            daemon_log(LOG_DEBUG, "Previous conflict packet received %llu ms  ago", usec);
+                            if (usec > (DEFEND_INTERVAL * 1000)) {
+                                daemon_log(LOG_DEBUG, "Clearing timestamp");
+                                clear_timestamp(&last_conflict);
+                            }
+                        }
+                        if ((should_defend) && (!has_timestamp(&last_conflict))) {
```

Comment:
@chombourger, could this logic be mis-used to great a network storm by intentionally sending conflicting ARP packets just *outside* the acceptable interval?  Or does that what the RFC intends?  Perhaps a defensive approach would be warranted, that prevented the previous block from clearing the timestamp and falling into this block, e.g. use an else here?  Perhaps the state machine here could be re-tooled?